ClassCastException when trying to convert a field from int to integer list

Hello,
I want to convert an existing int in my RealmObject to a RealmList containing a custom integer wrapper class. In my migration class I have some code to create the new list field, migrate the data from the old field to the new field and delete the old field. The code is somewhat adapted from [here](https://github.com/realm/realm-java/blob/master/examples/migrationExample/src/main/java/io/realm/examples/realmmigrationexample/model/Migration.java).

#### Goal

> Convert a int field to a realmlist containing integers in a custom wrapper class, as realm doesn't currently support `RealmList<Integer>`.

#### Expected Results

> I can convert the field and transform all the data from the old field to the new field using `transform`.

#### Actual Results

>  The app crashes as soon as I open Realm and the migration gets executed.

`Caused by: java.lang.ClassCastException: com.example.RealmInteger cannot be cast to io.realm.internal.RealmObjectProxy
                                                                        at io.realm.DynamicRealmObject.setList(DynamicRealmObject.java:602)
                                                                        at com.example.RealmHelper$Migration$1.apply(RealmHelper.java:132)
                                                                        at io.realm.RealmObjectSchema.transform(RealmObjectSchema.java:475)
                                                                        at com.example.RealmHelper$Migration.migrate(RealmHelper.java:126)
                                                                        at io.realm.BaseRealm$3.onResult(BaseRealm.java:705)
                                                                        at io.realm.RealmCache.invokeWithGlobalRefCount(RealmCache.java:267)
                                                                        at io.realm.BaseRealm.migrateRealm(BaseRealm.java:685)
                                                                        at io.realm.Realm.migrateRealm(Realm.java:1221)
                                                                        at io.realm.Realm.migrateRealm(Realm.java:1208)
                                                                        at io.realm.Realm.createInstance(Realm.java:235)
                                                                        at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:126)
                                                                        at io.realm.Realm.getDefaultInstance(Realm.java:174)
                                                                        at com.example.RealmHelper.init(RealmHelper.java:86)
                                                                        at com.example.AppApplication.onCreate(AppApplication.java:95)
                                                                        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1014)
                                                                        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4751)
                                                                        at de.robv.android.xposed.XposedBridge.invokeOriginalMethodNative(Native Method) 
                                                                        at de.robv.android.xposed.XposedBridge.handleHookedMethod(XposedBridge.java:729) 
                                                                        at android.app.ActivityThread.handleBindApplication(<Xposed>) 
                                                                        at android.app.ActivityThread.-wrap1(ActivityThread.java) 
                                                                        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1424) 
                                                                        at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                        at android.os.Looper.loop(Looper.java:148) 
                                                                        at android.app.ActivityThread.main(ActivityThread.java:5461) 
                                                                        at java.lang.reflect.Method.invoke(Native Method) 
                                                                        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
                                                                        at de.robv.android.xposed.XposedBridge.main(XposedBridge.java:133) `

#### Steps & Code to Reproduce

> Describe your current debugging efforts.

#### Code Sample


RealmInteger:

```java

public class RealmInteger extends RealmObject {

    private int value;

    public RealmInteger() {
    }

    public RealmInteger(int value) {
        this.value = value;
    }

    public int getValue() {
        return value;
    }

    public void setValue(int value) {
        this.value = value;
    }
}

```

Migration:

```java

RealmObjectSchema integerSchema = schema.create("RealmInteger")
        .addField("value", int.class);

RealmObjectSchema objectSchema = schema.get("CustomObject");

// Remove trip and add trips
objectSchema.addRealmListField("trips", integerSchema)
        .transform(new RealmObjectSchema.Function() {
            @Override
            public void apply(DynamicRealmObject object) {
                RealmList<RealmInteger> list = new RealmList<>();
                list.add(new RealmInteger(object.getInt("trip")));

                object.setList("trips", list);
            }
         })
         .removeField("trip");

```

Basically, I want to convert

```java

public class CustomObject extends RealmObject {
    private int trip;
}

```

to

```java

public class CustomObject extends RealmObject {
    private RealmList<RealmInteger> trips;
}

```

and possibly preserve the existing data (e.g. if `trip` contains `1`, `trips` should contain `[1]`) .

#### Version of Realm and tooling
Realm version(s): 1.0.0

Android Studio version: 2.2 Beta

Which Android version and device: 6.0.1 CM13 Galaxy S5

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,7 @@
 * `insert()` now correctly throws an exception if two different objects have the same primary key (#3212).
 * Blackberry Z10 throwing "Function not implemented" (#3178).
 * Reduced the number of file descriptors used by Realm Core (#3197).
+* `DynamicRealmObject.setList` should only accept `RealmList<DynamicRealmObject>` (#3280).
 
 ### Internal
 

--- a/realm/realm-library/src/main/java/io/realm/DynamicRealmObject.java
+++ b/realm/realm-library/src/main/java/io/realm/DynamicRealmObject.java
@@ -32,7 +32,9 @@ import io.realm.internal.android.JsonUtils;
  * Using a DynamicRealmObject is slower than using the regular RealmObject class.
  */
 public final class DynamicRealmObject extends RealmObject implements RealmObjectProxy {
+
     private final ProxyState proxyState = new ProxyState(this);
+
     /**
      * Creates a dynamic Realm object based on an existing object.
      *
@@ -405,7 +407,7 @@ public final class DynamicRealmObject extends RealmObject implements RealmObject
             setObject(fieldName, (DynamicRealmObject) value);
         } else if (valueClass == RealmList.class) {
             @SuppressWarnings("unchecked")
-            RealmList<RealmObject> list = (RealmList<RealmObject>) value;
+            RealmList<DynamicRealmObject> list = (RealmList<DynamicRealmObject>) value;
             setList(fieldName, list);
         } else {
             throw new IllegalArgumentException("Value is of an type not supported: " + value.getClass());
@@ -574,7 +576,7 @@ public final class DynamicRealmObject extends RealmObject implements RealmObject
      * of the object represented by the DynamicRealmObject doesn't match or any element in the list belongs to a
      * different Realm.
      */
-    public void setList(String fieldName, RealmList<? extends RealmModel> list) {
+    public void setList(String fieldName, RealmList<DynamicRealmObject> list) {
         if (list == null) {
             throw new IllegalArgumentException("Null values not allowed for lists");
         }
@@ -599,7 +601,7 @@ public final class DynamicRealmObject extends RealmObject implements RealmObject
         links.clear();
         Table linkTargetTable = links.getTargetTable();
         for (int i = 0; i < list.size(); i++) {
-            RealmObjectProxy obj = (RealmObjectProxy) list.get(i);
+            RealmObjectProxy obj = list.get(i);
             if (obj.realmGet$proxyState().getRealm$realm() != proxyState.getRealm$realm()) {
                 throw new IllegalArgumentException("Each element in 'list' must belong to the same Realm instance.");
             }

